### PR TITLE
[impeller] [vk] Implement device buffer support

### PR DESCRIPTION
**Pull request by iskakaushik**

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 1.
View them at https://flutter-engine-gold.skia.org/cl/github/34961

File: impeller/renderer/backend/vulkan/debug_context_vk.cc [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/renderer/backend/vulkan/debug_context_vk.h"
+#include "impeller/base/validation.h"
+#include "vulkan/vulkan_enums.hpp"
+#include "vulkan/vulkan_structs.hpp"
+
+namespace impeller {
+DebugContextVK::DebugContextVK(const vk::Device& logical_device)
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/debug_context_vk.cc:11]: Minor nit: Newline after namespace.

File: impeller/renderer/backend/vulkan/debug_context_vk.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  ~DebugContextVK();
+
+  template <typename T>
+  bool SetDebugName(T handle, std::string_view label) const {
+    uint64_t handle_ptr =
+        reinterpret_cast<uint64_t>(static_cast<typename T::NativeType>(handle));
+
+    std::string label_str = std::string(label);
+
+    auto ret = logical_device_.setDebugUtilsObjectNameEXT(
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/debug_context_vk.h:25]: In the GLES backend, debug object labelling is [handled by the reactor](https://github.com/flutter/engine/blob/9b278ae58fbe9e94a174ea3dd493960103f5a400/impeller/renderer/backend/gles/reactor_gles.h#L51). Perhaps we can do something similar in the Vulkan backend and move this call to the `ContextVK`. A context is always available and it always has a device handle ready.

IMO that would be a lot simpler and keeping in line with the other API backends.

File: impeller/renderer/backend/vulkan/device_buffer_vk.h [hunk 3]
```diff
@@ -6,20 +6,47 @@
...
+  DeviceBufferAllocationVK(const VmaAllocator& allocator,
+                           VkBuffer buffer,
+                           VmaAllocation allocation,
+                           VmaAllocationInfo allocation_info);
+
+  ~DeviceBufferAllocationVK();
+
+  vk::Buffer GetBufferHandle() const;
+
+  void* GetMappedPtr() const;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/device_buffer_vk.h:25]: Minor nit: `GetBuffer` of `GetMapping` instead?

File: impeller/renderer/backend/vulkan/device_buffer_vk.cc [hunk 4]
```diff
@@ -3,9 +3,78 @@
...
+    VkBuffer buffer,
+    VmaAllocation allocation,
+    VmaAllocationInfo allocation_info)
+    : allocator_(allocator),
+      buffer_(buffer),
+      allocation_(allocation),
+      allocation_info_(allocation_info) {}
+
+DeviceBufferAllocationVK::~DeviceBufferAllocationVK() {
+  vmaDestroyBuffer(allocator_, buffer_, allocation_);
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/device_buffer_vk.cc:21]: Checking here for null arguments or return values (if there is one)?

File: impeller/renderer/backend/vulkan/device_buffer_vk.cc [hunk 5]
```diff
@@ -3,9 +3,78 @@
 // found in the LICENSE file.
 
 #include "impeller/renderer/backend/vulkan/device_buffer_vk.h"
+#include "fml/logging.h"
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/device_buffer_vk.cc:6]: Minor nit: Newline between the primary header and other includes.

File: impeller/renderer/backend/vulkan/device_buffer_vk.h [hunk 6]
```diff
@@ -6,20 +6,47 @@
...
+
+ private:
+  const VmaAllocator& allocator_;
+  vk::Buffer buffer_;
+  VmaAllocation allocation_;
+  VmaAllocationInfo allocation_info_;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/device_buffer_vk.h:31]: Here and elsewhere, `FML_DISALLOW_COPY_AND_ASSIGN`

File: impeller/renderer/backend/vulkan/device_buffer_vk.h [hunk 7]
```diff
@@ -6,20 +6,47 @@
...
   // |DeviceBuffer|
   ~DeviceBufferVK() override;
 
  private:
-  friend class AllocatorVK;
+  const DebugContextVK& debug_context_;
+  std::unique_ptr<DeviceBufferAllocationVK> device_allocation_;
 
-  DeviceBufferVK(size_t size, StorageMode mode);
+  friend class AllocatorVK;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/device_buffer_vk.h:49]: Minor nit: The convention in the other TUs is to put this after the `private:` label that needs needs the friend call.

File: impeller/renderer/backend/vulkan/device_buffer_vk.h (hunk 6 above)
Comment (iskakaushik) [reply impeller/renderer/backend/vulkan/device_buffer_vk.h:33]: Done!

File: impeller/renderer/backend/vulkan/device_buffer_vk.h (hunk 7 above)
Comment (iskakaushik) [reply impeller/renderer/backend/vulkan/device_buffer_vk.h:50]: Done!

File: impeller/renderer/backend/vulkan/device_buffer_vk.h (hunk 3 above)
Comment (iskakaushik) [reply impeller/renderer/backend/vulkan/device_buffer_vk.h]: Done!

File: impeller/renderer/backend/vulkan/device_buffer_vk.cc (hunk 4 above)
Comment (iskakaushik) [reply impeller/renderer/backend/vulkan/device_buffer_vk.cc]: Done! This doesn't have a return value.

File: impeller/renderer/backend/vulkan/device_buffer_vk.cc (hunk 5 above)
Comment (iskakaushik) [reply impeller/renderer/backend/vulkan/device_buffer_vk.cc:7]: Done!

File: impeller/renderer/backend/vulkan/debug_context_vk.h (hunk 2 above)
Comment (iskakaushik) [reply impeller/renderer/backend/vulkan/debug_context_vk.h]: This is resolved as well.

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**